runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/grpcclient/GrpcGetWorkerMetadataStream.java

```diff
@@ -0,0 +1,285 @@
+   * exposing a static method so that an instance of {@link GrpcGetWorkerMetadataStream} will be
+   * created in the correct state.
+   */
+  private void fullyInitializeRequestObserver() {
+    getWorkMetadataStreamRequestObserver.init(
+        AbstractWindmillStream.ResponseObserver::new,
+        () -> {
+          synchronized (GrpcGetWorkerMetadataStream.this) {
+            streamClosed.set(true);
+          }
+        });
+  }
+
+  @Override
+  protected StreamObserver<WorkerMetadataRequest> createNewRequestObserver() {
+    // we do not need to re-initiate the requestObserver here, since every
+    // GetWorkMetadataStreamRequestObserver.onNext() starts a new server stream.
+    return getWorkMetadataStreamRequestObserver;
+  }
+
+  @Override
+  protected void onResponse(WorkerMetadataResponse response) {
+    synchronized (this) {
+      // If the metadataVersion newer or the same as the one in the response, the response data is
+      // stale, and we do not want to do anything.
+      long currentMetadataVersion = metadataVersion.get();
+      if (currentMetadataVersion >= response.getMetadataVersion()) {
+        LOG.info(
+            "Received metadata version={}; Current metadata version={}. "
+                + "Skipping update because received stale metadata",
+            response.getMetadataVersion(),
+            currentMetadataVersion);
+        return;
+      }
+      metadataVersion.set(response.getMetadataVersion());
+    }
+    ImmutableMap<String, String> globalDataServers =
+        response.getGlobalDataEndpointsMap().entrySet().stream()
+            .collect(
+                toImmutableMap(
+                    Map.Entry::getKey, // global data key
+                    endpoint -> endpoint.getValue().getEndpoint()));
+
+    ImmutableList<String> windmillServers =
+        response.getDirectPathEndpointsList().stream()
+            .map(WorkerMetadataResponse.Endpoint::getEndpoint)
+            .collect(toImmutableList());
+
+    serverMappingUpdater.accept(
+        WindmillEndpoints.builder()
+            .setGlobalDataServers(globalDataServers)
+            .setWindmillServers(windmillServers)
+            .build());
+  }
+
+  @Override
+  protected synchronized void onNewStream() {
+    send(workerMetadataRequest);
+  }
+
+  @Override
+  protected void startThrottleTimer() {
+    getWorkerMetadataThrottleTimer.start();
+  }
+
+  @Override
+  protected void appendSpecificHtml(PrintWriter writer) {
+    writer.format(
+        "GetWorkerMetadataStream: version=[%d] , job_header=[%s]",
+        getMetadataVersion(), workerMetadataRequest.getHeader());
+  }
+
+  @Override
+  protected StreamObserver<WorkerMetadataRequest> requestObserver() {
+    return getWorkMetadataStreamRequestObserver;
+  }
+
+  private synchronized long getMetadataVersion() {
+    return metadataVersion.get();
+  }
+
+  /**
+   * GetWorkerMetadata is a server streaming API. There is no request observer since the client side
```

Comment:
It seems that a lot of the changes here are necessary because this is not a bi-directional stream.

Beyond the code complexity of supporting this, I'm concerned about us not being able to send application-level heartbeat requests as we've had numerous issues in the past where sending was necessary to detect that the stream had died. See https://issues.apache.org/jira/browse/BEAM-10808 and https://github.com/apache/beam/pull/12688.

I'm thinking that we should change the metadata streaming rpc to be bi-directional instead of just server side streaming to make sure we don't hit similar issues.

Reply:
done.